// canvas-toolkit/build/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _elementsLine = require('./elements/Line');

var _elementsLine2 = _interopRequireDefault(_elementsLine);

var _elementsPath = require('./elements/Path');

var _elementsPath2 = _interopRequireDefault(_elementsPath);

var _elementsRectangle = require('./elements/Rectangle');

var _elementsRectangle2 = _interopRequireDefault(_elementsRectangle);

var _elementsText = require('./elements/Text');

var _elementsText2 = _interopRequireDefault(_elementsText);

var _elementsCircle = require('./elements/Circle');

var _elementsCircle2 = _interopRequireDefault(_elementsCircle);

var _elementsTriangle = require('./elements/Triangle');

var _elementsTriangle2 = _interopRequireDefault(_elementsTriangle);

var _coreCanvas = require('./core/Canvas');

var _coreCanvas2 = _interopRequireDefault(_coreCanvas);

var _coreCanvasElement = require('./core/CanvasElement');

var _coreCanvasElement2 = _interopRequireDefault(_coreCanvasElement);

var _coreLayeredElement = require('./core/LayeredElement');

var _coreLayeredElement2 = _interopRequireDefault(_coreLayeredElement);

var _coreUtil = require('./core/util');

var CanvasToolkit = {
  Path: _elementsPath2['default'],
  Rectangle: _elementsRectangle2['default'],
  Triangle: _elementsTriangle2['default'],
  Text: _elementsText2['default'],
  Line: _elementsLine2['default'],
  Circle: _elementsCircle2['default'],
  Canvas: _coreCanvas2['default'],
  CanvasElement: _coreCanvasElement2['default'],
  LayeredElement: _coreLayeredElement2['default'],
  RedrawProperties: _coreUtil.RedrawProperties
};

window.CanvasToolkit = CanvasToolkit;

exports['default'] = CanvasToolkit;
module.exports = exports['default'];
