// canvas-toolkit/build/core/Canvas.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _underscore = require('underscore');

var _underscore2 = _interopRequireDefault(_underscore);

var _Layer = require('./Layer');

var _Layer2 = _interopRequireDefault(_Layer);

var _LayeredElement = require('./LayeredElement');

var _LayeredElement2 = _interopRequireDefault(_LayeredElement);

var _CanvasElement = require('./CanvasElement');

var _CanvasElement2 = _interopRequireDefault(_CanvasElement);

var Canvas = (function () {
  function Canvas(_ref) {
    var _this = this;

    var width = _ref.width;
    var height = _ref.height;
    var container = _ref.container;

    _classCallCheck(this, Canvas);

    this.container = container;
    this.composites = [];
    this.defaultLayer = new _Layer2['default']({ width: width, height: height });
    this.container.appendChild(this.defaultLayer.canvasEl);
    this.width = width;
    this.height = height;

    var drawLoop = function drawLoop() {
      _this.defaultLayer.draw();
      _underscore2['default'].each(_this.composites, function (composite) {
        composite.draw();
      });
      window.requestAnimationFrame(drawLoop);
    };

    window.requestAnimationFrame(drawLoop);
  }

  _createClass(Canvas, [{
    key: 'clear',
    value: function clear() {
      this.defaultLayer.clear();
      this.composites = [];
      while (this.container.firstChild) {
        this.container.removeChild(this.container.firstChild);
      }
      this.container.appendChild(this.defaultLayer.canvasEl);
    }
  }, {
    key: 'add',
    value: function add(element) {
      var _this2 = this;

      if (element instanceof _LayeredElement2['default']) {
        element.width = this.width;
        element.height = this.height;
        this.composites.push(element);
        _underscore2['default'].each(element.layers, function (layer) {
          _this2.container.appendChild(layer.canvasEl);
        });
      } else if (element instanceof _CanvasElement2['default']) {
        this.defaultLayer.add(element);
      } else {
        throw 'Unrecognized Element';
      }
    }
  }, {
    key: 'width',
    get: function () {
      return this._width;
    },
    set: function (val) {
      this._width = val;
      _underscore2['default'].each(this.composites, function (composite) {
        composite.width = val;
      });
      this.defaultLayer.width = val;
    }
  }, {
    key: 'height',
    get: function () {
      return this._height;
    },
    set: function (val) {
      this._height = val;
      _underscore2['default'].each(this.composites, function (composite) {
        composite.height = val;
      });
      this.defaultLayer.height = val;
    }
  }]);

  return Canvas;
})();

exports['default'] = Canvas;
module.exports = exports['default'];
