// canvas-toolkit/build/core/LayeredElement.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _underscore = require('underscore');

var _underscore2 = _interopRequireDefault(_underscore);

var _Layer = require('./Layer');

var _Layer2 = _interopRequireDefault(_Layer);

var LayeredElement = (function () {
  function LayeredElement() {
    _classCallCheck(this, LayeredElement);

    this.layers = [];
  }

  _createClass(LayeredElement, [{
    key: 'draw',
    value: function draw() {
      _underscore2['default'].each(this.layers, function (layer) {
        layer.draw();
      });
    }
  }, {
    key: 'createLayer',
    value: function createLayer() {
      var layer = new _Layer2['default']({
        width: this.width,
        height: this.height
      });
      this.layers.push(layer);
      return layer;
    }
  }, {
    key: 'width',
    get: function () {
      return this._width;
    },
    set: function (val) {
      this._width = val;
      _underscore2['default'].each(this.layers, function (layer) {
        layer.width = val;
      });
    }
  }, {
    key: 'height',
    get: function () {
      return this._height;
    },
    set: function (val) {
      this._height = val;
      _underscore2['default'].each(this.layers, function (layer) {
        layer.height = val;
      });
    }
  }]);

  return LayeredElement;
})();

exports['default'] = LayeredElement;
module.exports = exports['default'];
